app/endpoints/raffle.py

```diff
@@ -0,0 +1,738 @@
+    """
+
+    typeticket = await cruds_raffle.get_typeticket_by_id(
+        typeticket_id=typeticket_id, db=db
+    )
+    if not typeticket:
+        raise HTTPException(status_code=404, detail="Typeticket not found")
+
+    await cruds_raffle.delete_typeticket(typeticket_id=typeticket_id, db=db)
+
+
+@router.get(
+    "/tombola/raffle/{raffle_id}/type_tickets",
+    response_model=list[schemas_raffle.TypeTicketSimple],
+    status_code=200,
+    tags=[Tags.raffle],
+)
+async def get_type_tickets_by_raffle_id(
+    raffle_id: str,
+    db: AsyncSession = Depends(get_db),
+    user: models_core.CoreUser = Depends(is_user_a_member),
+):
+    """
+    Return all type_tickets associated to a raffle
+    """
+    type_tickets = await cruds_raffle.get_typeticket_by_raffleid(raffle_id, db)
+    return type_tickets
+
+
+@router.get(
+    "/tombola/tickets",
+    response_model=list[schemas_raffle.TicketSimple],
+    status_code=200,
+    tags=[Tags.raffle],
+)
+async def get_tickets(
+    db: AsyncSession = Depends(get_db),
+    user: models_core.CoreUser = Depends(is_user_a_member_of(GroupType.soli)),
+):
+    """
+    Return all tickets
+
+    **The user must be a member of the group soli to use this endpoint**
+    """
+    tickets = await cruds_raffle.get_tickets(db)
+    return tickets
+
+
+@router.post(
+    "/tombola/tickets/buy/{type_id}",
+    response_model=list[schemas_raffle.TicketComplete],
+    status_code=201,
+    tags=[Tags.raffle],
+)
+async def buy_ticket(
+    type_id: str,
+    db: AsyncSession = Depends(get_db),
+    redis_client: Redis | None = Depends(get_redis_client),
+    user: models_core.CoreUser = Depends(is_user_a_member),
+    request_id: str = Depends(get_request_id),
+):
+    """
+    Buy a ticket
+    """
+    type_ticket = await cruds_raffle.get_typeticket_by_id(typeticket_id=type_id, db=db)
+    if type_ticket is None:
+        raise ValueError("Bad typeticket association")
+
+    balance: models_raffle.Cash | None = await cruds_raffle.get_cash_by_id(
+        db=db,
+        user_id=user.id,
+    )
+
+    # If the balance does not exist, we create a new one with a balance of 0
+    if not balance:
+        new_cash_db = schemas_raffle.CashDB(
+            balance=0,
+            user_id=user.id,
+        )
+        balance = models_raffle.Cash(
+            **new_cash_db.dict(),
+        )
+        await cruds_raffle.create_cash_of_user(
+            cash=balance,
+            db=db,
+        )
+    db_ticket = [
+        models_raffle.Tickets(id=str(uuid.uuid4()), type_id=type_id, user_id=user.id)
+        for i in range(type_ticket.pack_size)
+    ]
+
+    for ticket in db_ticket:
+        ticket.user = user
+        ticket.type_ticket = type_ticket
+
+    redis_key = "raffle_" + user.id
+
+    if not isinstance(redis_client, Redis) or locker_get(
+        redis_client=redis_client, key=redis_key
+    ):
+        raise HTTPException(status_code=403, detail="Too fast !")
```

Comment:
403 is reserved for Unauthorized request, at the auth level. A 429 error code may be used.
https://developer.mozilla.org/fr/docs/Web/HTTP/Status/429

Reply:
Fixed